### ISSUE TITLE
Plan for adding a CUDA backend to the DNN module

### ISSUE THREAD

**Issue by YashasSamaga** [opencv/opencv#14585]
##### Detailed description

As a part of GSoC work, I will be writing a CUDA backend for the DNN module. I have come up with a plan which requires technical review.

The project can be broadly divided into two components: CUDA Support Library (CSL) and the actual layer implementation. The CSL provides a neat high-level C++ interface and abstracts away the CUDA, cuDNN and cuBLAS API (which are actually C APIs). The layer implementations will be high-level code which work with components from the CSL such as `csl::Tensor`.

The project requires adding cuDNN as a new dependency which will be controlled by a new `WITH_CUDNN` option. A new option `OPENCV_DNN_CUDA` (similar to `OPENCV_DNN_OPENCL`) will be added which requires `WITH_CUDA`, `WITH_CUBLAS`, `WITH_CUDNN`. If the requirements are met, `CV_CUDA4DNN` (similar to `CV_OCL4DNN`) preprocessor symbol will be added which will be used to conditionally compile CUDA code.

New targets `DNN_TARGET_CUDA_[FP16/FP32/FP64]` will be added to `DNN_BACKEND_OPENCV`. The layers will be implemented inside the existing layer classes similar to what the OpenCL backend currently does. This involves adding a few new members mainly tensor objects and two new methods `forwardCUDA` and `initCUDA`. `initCUDA` is about converting the blobs to CSL tensors and setting up a few other things.

A far more detailed plan can be found here: 
https://docs.google.com/document/d/1gJ2Pc8XZUaFEGW5LOx-MF9lfWvmR3mjqht0t_bufIA8/edit?usp=sharing

There is also a [prototype ](https://github.com/YashasSamaga/cuda-dnn-inference) which vaguly shows the division of work into CSL and the actual layer implementation.

**PR Schedule:**

1. CMake Update PR 1
2. CSL PR 1: adds the CSL base and components required for fc, softmax and activation layer
3. Stub PR: sets up all layers for CUDA target which fallback to CPU
4. FC PR (to perform FC forward pass on GPU)
5. Activation PR
6. Softmax PR
7. TBD (begining of June)

**Comment by davisking** [comment]
I think this all sounds good (I'm also @YashasSamaga's mentor for GSoC).  I'm not an OpenCV maintainer though so one of them should chime in if there are objections or things that a maintainer wants clarified. But based on what I've seen from @YashasSamaga so far I think he is on the right track :)

**Comment by YashasSamaga** [comment]
I particularly wanted to know what the consensus was for:
- adding new virtual methods, `initCUDA` and `forwardCUDA` , to the `Layer` class (breaks ABI)
- not using the `GpuMat` and barely using the existing CUDA code

---

**Plan 1 (current):**
It requires adding two _new virtual methods_ — `initCUDA` and `forwardCUDA` — to the `Layer` class. `forwardCUDA`will by default invoke `forward` method. This allows a user to write a CPU version of a custom layer and still use CUDA by allowing the missing CUDA implementation to fallback to CPU. If the user wants to write a CUDA implementation, it's reasonably easy. All they would have to do is add code to `initCUDA` to store parameters in GPU tensors and then override the `forwardCUDA` method.  

- easier to write custom layers
- breaks ABI
- inconsistent
  * `init<Backend>` is used for initializing backends but `initCUDA` will be initializing stuff for a target for `DNN_BACKEND_OPENCV`

**Plan 2:**
Add `CUDABackendNode` which has `initCUDA` and `forwardCUDA` along with CUDA specific member objects. A class for every layer, `CUDAFullyConnectedLayer` for example, will be created which inherit from `BackendNode`. These new classes will be added alongside the CPU implementations in the existing layer files.

- no ABI change
- makes the process of writing custom layers for GPUs a bit complex
- inherits `BackendNode` for `DNN_BACKEND_OPENCV` which isn't what it was meant to be used as (at least according to this comment  ` // Computation nodes of implemented backends (except DEFAULT).
`)

**Plan 3:**
Use `GpuMat` to store a memory block and obtain it from `InputArrayOfArrays/OutputArrayOfArrays` in `forward` method. Convert it to `csl::Tensor`. Use lazy initialization to invoke `initCUDA` in `forward` method.

- can also pass a `void*` through `InputArrayOfArrays/OutputArrayOfArrays`

These sound more like hacks to get around a different problem.

**Plan 4:**
Add non-virtual `forwardCUDA` and `initCUDA` in every layer class. Perform dynamic cast from `Layer` to the actual layer class and invoke the methods.

How to perform the dynamic cast?
- a huge try-catch block until we hit the correct cast
- get type and directy cast using a switch-case (more or less the same as the previous one)

Abuse of RTTI? Also requires extra work to add a new layer.

**Comment by dkurt** [comment]
Hi, @YashasSamaga and @davisking!

We discussed recently with our team that it'd be nice to refactor dnn backends to make it modular. As for now I think it's OK to follow the **Plan 1**. From our experience with other backends there are similar entities for each of them: buffer and node.
| | buffer | node |
|---|---|---|
|OpenCL | UMat | kernel |
| IE | wrapper over Mat (IE's blob) | `InferenceEngine::Builder::Network` (which is a set of `InferenceEngine::Builder::Layer`) |
|  Halide | Halide::Buffer | Halide::Func |

`BackendWrapper` and `BackendNode` are used to store buffers and nodes correspondingly.

So I'd like to suggest to follow the simplest way for now.

**Comment by alalek** [comment]
> DNN_TARGET_CUDA_[FP16/FP32/FP64] will be added to DNN_BACKEND_OPENCV

CUDNN is external accelerator, so it make sense to create new backend:
- `DNN_BACKEND_CUDA`
- `DNN_TARGET_CUDA_[FP16/FP32/FP64]` (not sure if we need FP64 for now)

---

<details>

<summary>Some other thoughts...</summary>

Regarding targets, it make sense create `_GPU_` entries (and merge them with existed OpenCL):
```
DNN_TARGET_GPU_FP32 = DNN_TARGET_OPENCL,
DNN_TARGET_GPU_FP16 = DNN_TARGET_OPENCL_FP16,
DNN_TARGET_GPU_FP64,
```

Using these statements:
- Backend is optimization technology
- Target is mostly related to H/W device (CPU, GPU, Myriad accelerator, FPGA accelerator)
- Precision looks like a hack for now. Some networks has mixed precision for different layers. Especially some minimized versions like INT8, INT1.

But...
- `VULKAN` doesn't fit
- Halide + [OpenCL/CUDA]
- OpenCL + [CPU device]

</summary>

**Comment by dkurt** [comment]
@alalek, maybe even `DNN_BACKEND_CUDNN`? Because as far as I understood this library will be used for, not raw CUDA.

**Comment by YashasSamaga** [comment]
FP64 comes for free after FP32 is implemented. FP16 requires a little bit of work.

Adding a backend entry simplifies the code as it nicely fits into the existing code. Currently, in attempts to use `DNN_BACKEND_OPENCV`, I have been adding additional branches all over the place to handle the special case.

`DNN_BACKEND_OPENCV` with `DNN_TARGET_CUDA_FPXX`:
--- 
The `initCUDA` and `forwardCUDA` would be added to the user-facing `Layer` class. The `initCUDA` defaults to doing nothing when not implemented and `forwardCUDA` throws an exception which is caught by `forwardLayer` (dnn.cpp) and causes it to switch to `DNN_TARGET_CPU` for that layer.

When users implement a custom layer, they have the choice to provide a CUDA implementation for it. If they do not provide a CUDA implementation, the layer will fallback to the CPU version. If they do provide a CUDA implementation, it will be used.

**Why `DNN_BACKEND_OPENCV`?**
- `forwardCUDA` method is added to the `Layer` class
- vast majority of the code will be in raw CUDA (and with the help of cuBLAS )

**Problems:**
- (`DNN_BACKEND_OPENCV`, `DNN_TARGET_CPU`) does not support FP16/FP64; any fallback from FP16/FP64 versions will completely fail (FP16 can be salvaged by converting it to FP32 and then converting back to FP16)

**Advantages:**
- custom layers are supported (which requires exposing the CSL components to the user)

**Disadvantages:**
- the implementation code becomes a bit complex as it involves adding special cases under `DNN_BACKEND_OPENCV`

`DNN_BACKEND_XXXX` with `DNN_TARGET_CUDA_FPXX`:
---
If custom CUDA layers support isn't required, there is no need to add `initCUDA` and `forwardCUDA` to  the `Layer` class. Everything will be dealt using `BackendNode` derivatives and the only change to user-facing API would be adding `initXXXXBackend` to the `Layer` class just like what IE and Halide backend do.

**Advantages:**
- simplifies the code
- easier to understand as it follows the code pattern followed by IE or Halide backend
- CSL is completely an implementation detail and won't be exposed externally

**Disadvantages:**
- no custom CUDA layers

cuDNN has limited capabilities
---

cuDNN provides only a small subset of layers:
1. convolution (and deconvolution)
2. pooling (max and average)
3. softmax (and log softmax)
4. ReLU, PReLU, clipped ReLU, TanH, Sigmoid
6. BN/LRN
7. LSTM

The remaining layers will be implemented using cuBLAS and raw CUDA.

**Comment by YashasSamaga** [comment]
@dkurt @alalek any final verdict?

**Comment by dkurt** [comment]
@YashasSamaga, I'd like to recommend to follow the simplest way. Anyway we'll review the changes :smile:

**Comment by YashasSamaga** [comment]
stub commit https://github.com/YashasSamaga/opencv/commit/76350b8bde2beddab2e38e4a6debea4fbd76e128

**Changes:**
- adds `DNN_BACKEND_CUDA` and `DNN_TARGET_CUDA_FP32`
- adds `initCUDA` and `forwardCUDA` methods to `Layer` with fallbacks

**Tests:**
- tested with a random LeNet model
- tested with `res10_300x300_ssd_iter_140000.caffemodel`

The fallbacks work correctly and the detection outputs are correct.

I'll make a PR once #14660 is merged.

**Comment by dkurt** [comment]
@YashasSamaga, I think it's better to make first PR with both dependencies and at least a single network so we can validate accuracy as well.

**Comment by YashasSamaga** [comment]
@dkurt I'm sorry. I could not understand your previous comment.

cuDNN is the only new dependency for the project which is taken care by the first PR.

The [stub commit](https://github.com/YashasSamaga/opencv/commit/76350b8bde2beddab2e38e4a6debea4fbd76e128) (on my fork) I previously linked stubs out the functionality. Even if CUDA backend is requested, the layers will fallback to using (`DNN_BACKEND_OPENCV`, `DNN_TARGET_CPU`) as CUDA implementations have not been added yet. I will adding a CUDA implementations to the layers one by one slowly.

Do you want me to add the stub commit to the first PR? Or make another PR? Or wait for the first PR to be merged?

The code I had used for testing can be found here: https://gist.github.com/YashasSamaga/4de2b9f96289bbbb868fc6eb90d2f1b4

The accuracy of the model on MNIST test dataset with the CUDA backend exactly matches the accuracy on the OpenCV backend with CPU target (which isn't surprising given that all layers fallback to the CPU versions even if CUDA backend was requested).

The MNIST model is a simple CNN. I also manually tested `res10_300x300_ssd_iter_140000.caffemodel` by giving a dozen sample images with and without faces. It detected the number of faces correctly.

**Comment by alalek** [comment]
> Or make another PR

This one.

New PR should contain commits from the pending PR (until it got merged - GitHub/Git will handle this automatically).
Just add a note "Merge after PR: #<number>" into description.
